### PR TITLE
bpo-39573: Convert Py_REFCNT and Py_SIZE to functions

### PR DESCRIPTION
**Pull request by vstinner**
Convert Py_REFCNT() and Py_SIZE() macros to static inline functions.
They cannot be used as l-value anymore: use Py_SET_REFCNT() and
Py_SET_SIZE() to set an object reference count and size.

Replace &Py_SIZE(self) with &((PyVarObject*)self)->ob_size
in arraymodule.c.

This change is backward incompatible on purpose, to prepare the C API
for a fully opaque PyObject structure.

https://bugs.python.org/issue39573

### CODE REVIEW

**Review by corona10 — COMMENTED** [review]
```
Objects/tupleobject.c:85:21: error: expression is not assignable
        Py_SIZE(op) = size;
        ~~~~~~~~~~~ ^
```

**Review by corona10 — CHANGES_REQUESTED** [review]
I left comment ;)

**Comment by vstinner** [comment]
> Objects/tupleobject.c:85:21: error: expression is not assignable

Oh, Py_TRACE_REFS! Thanks for testing for me. It's now fixed.

**Review by corona10 — APPROVED** [review]
lgtm

**Comment by vstinner** [comment]
This PR breaks Cython. I proposed https://github.com/cython/cython/pull/3639 to use Py_SET_SIZE(): PR for the Cython master branch (then I will have to backport it to the 0.29.x branch). Maybe we should wait for a new Cython release compatible with this PR, before merging this PR.

**Comment by vstinner** [comment]
> This PR breaks Cython. I proposed cython/cython#3639 to use Py_SET_SIZE(): PR for the Cython master branch (then I will have to backport it to the 0.29.x branch).

My fix landed in 0.29.x and master branches of Cython, great!

I asked https://github.com/cython/cython/pull/3639#issuecomment-634636316 if a Cython 0.29.20 version will be released with the fix.

In the meanwhile, I think that it's fine to break Cython for a few days/weeks since we are at the beginning of the Python 3.10 dev cycle. So I merge my PR to be able to move on other C API issues.

**Pull request merged**
